include/spotify/json/codec/string.hpp

```diff
@@ -115,14 +115,36 @@ class string_t final {
     detail::fail(context, "\\u must be followed by 4 hex digits");
   }
 
-  static void decode_unicode_escape(decode_context &context, std::string &out) {
+  static void decode_unicode_escape(decode_context &context,
+                                    std::string &out,
+                                    unsigned high_surrogate) {
     detail::require_bytes<4>(context, "\\u must be followed by 4 hex digits");
     const auto a = decode_hex_nibble(context, *(context.position++));
     const auto b = decode_hex_nibble(context, *(context.position++));
     const auto c = decode_hex_nibble(context, *(context.position++));
     const auto d = decode_hex_nibble(context, *(context.position++));
     const auto p = unsigned((a << 12) | (b << 8) | (c << 4) | d);
-    encode_utf8(out, p);
+    if (json_unlikely(high_surrogate != 0)) {
+      if (json_likely(is_low_surrogate(p))) {
+        // Any Unicode codepoint encoded by a surrogate pair is 4 bytes in UTF-8
+        encode_utf8_4(out, codepoint_from_surrogate_pair(high_surrogate, p));
+      } else {
+        detail::fail(context, "Codepoint following high surrogate is not a low surrogate");
+      }
+    } else if (json_unlikely(is_high_surrogate(p))) {
+      // Parse low surrogate
```

Comment:
As discussed offline, I'll add an overload for `skip_1` that takes a custom error message instead.